### PR TITLE
Feature/parametric pulses

### PR DESCRIPTION
**Pull request by fr34q**
Pull request to track changes relative to master branch.

**Feature ist still in development so do not merge!**

### CODE REVIEW

File: qkit/measure/timedomain/pulse_sequence.py [hunk 1]
```diff
@@ -46,73 +63,275 @@ def __call__(self, time_fractions):
...
+        """
+        pulse_dict = {}
+        if pulse.name in ["readout", "wait"]:
+            print(pulse.name + " is not permitted as pulse name.")
+            print("Pulse name set to None.")
+            pulse_dict["name"] = None
+        else:
+            pulse_dict["name"] = pulse.name
+        pulse_dict["shape"] = pulse.shape.name
+        if isinstance(pulse.length, float):
```
Comment (fr34q) [inline qkit/measure/timedomain/pulse_sequence.py]: what if int/long ;) (okay, not very plausible)

File: qkit/measure/timedomain/pulse_sequence.py [hunk 2]
```diff
@@ -46,73 +63,275 @@ def __call__(self, time_fractions):
...
 
-            if not isinstance(pulse, Pulse):
-                # Only a marker pulse -> not visible in waveform
-                continue
+    def add_wait(self, time):
+        """
+        Add a wait time to the sequence.
+        Using a (lambda) function makes the wait time variable.
+        """
+        pulse_dict = {}
```
Comment (fr34q) [inline qkit/measure/timedomain/pulse_sequence.py]: Code duplicity with add()...
Maybe move common code to extra private method
like  `_add_pulse(name, length, skip=False)`

File: qkit/measure/timedomain/pulse_sequence.py [hunk 3]
```diff
@@ -46,73 +63,275 @@ def __call__(self, time_fractions):
...
+    def get_pulses(self):
+        """
+        Returns a list of all pulses and their properties. 
+        The properties of each pulse are stored in a dictionary, containing its name, shape, length and skip parameter.
+        """
+        dict_list = []
+        for pulse_dict in self._pulses:
+            temp = {}
+            temp["name"] = pulse_dict["name"]
+            if temp["name"] not in ["readout", "wait"]:
```
Comment (fr34q) [inline qkit/measure/timedomain/pulse_sequence.py]: Fallunterscheidungen wirklich nötig?

Comment (fr34q) [reply qkit/measure/timedomain/pulse_sequence.py]: Könnten sonst einfach self._pulses zurückgeben

File: qkit/measure/timedomain/pulse_sequence.py [hunk 4]
```diff
@@ -46,73 +63,275 @@ def __call__(self, time_fractions):
...
+        # hide y ticks
+        ax.set_yticks([])
+        # hide top and right spines
+        ax.spines['right'].set_visible(False)
+        ax.spines['top'].set_visible(False)
+        return
 
-        return waveform
+    def _pulselength_as_str(self, pulse_length):
+        # Note: currently only works for lambda functions.
```
Comment (fr34q) [inline qkit/measure/timedomain/pulse_sequence.py]: for functions one could use the function name instead?

File: qkit/measure/timedomain/sequence_library.py [hunk 5]
```diff
@@ -0,0 +1,122 @@
...
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program; if not, write to the Free Software
+# Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
+
+import pulse_sequence as ps
+
+
+def compensate():
```
Comment (fr34q) [inline qkit/measure/timedomain/sequence_library.py]: Empty function?

File: qkit/measure/timedomain/pulse_sequence.py [hunk 6]
```diff
@@ -1,39 +1,56 @@
...
-        self.length = length  # type: float
-        self.shape = shape  # type: Shape
+    def __init__(self, length, shape=ShapeLib.rect, name = None, amplitude=1, frequency_shift=0):
+        """
+        Input:
+            length          - length of the pulse. This can also be a (lambda) function for variable pulse lengths.
+            shape           - pulse shape (i.e. rect, gauss, ...)
+            name            - name you want to give your pulse (i.e. pi-pulse, ...)
+            amplitude       - relative amplitude of your pulse
+            frequency_shift - frequency shift of your pulse (currently not implemented)
```
Comment (fr34q) [inline qkit/measure/timedomain/pulse_sequence.py]: What would that do again?

File: qkit/measure/timedomain/sequence_library.py [hunk 7]
```diff
@@ -0,0 +1,122 @@
...
+            sample          - sample object.
+            pulse_shape     - Shape of the pulses (i.e. square, gauss, ...)
+            amplitude       - relative ampitude of the sequence.
+            frequency_shift - frequency shift of the pulses.
+
+        Output:
+            Sequence object for rabi experiment.
+    """
+    rabi_tone = ps.Pulse(lambda t: t, shape = pulse_shape, name = "rabi-tone", amplitude = amplitude, frequency_shift = frequency_shift)
+    sequence = ps.PulseSequence(sample, sample.clock)
```
Comment (fr34q) [inline qkit/measure/timedomain/sequence_library.py]: passing `sample.clock` is superfluous

The same applies for all experiments below

File: qkit/measure/timedomain/TdExperiment.py [hunk 8]
```diff
@@ -0,0 +1,440 @@
...
+            seq_dict["sequence_%i"%i] = self._sequences[i].get_pulses()
+            seq_dict["time_%i"%i] = self._times[i]
+        seq_dict["interleave"] = self._interleave
+        return seq_dict
+
+    def interleave(self, value = True):
+        """
+        Sequences with an equal number of timesteps may be interleaved.
+        """
+        if value is False:
```
[the remark this answers is not in the document]
Comment (AlexStehli) [reply qkit/measure/timedomain/TdExperiment.py]: Fixed, now also sets interleave to False if value is False.

File: qkit/measure/timedomain/TdExperiment.py [hunk 9]
```diff
@@ -0,0 +1,440 @@
...
+        # Dictionary for x-axis scaling
+        self._x_unit = { "s": 1, "ms": 1e-3, "us": 1e-6, "ns": 1e-9}
+
+    def add_sequence(self, sequence, *args):
+        """
+        Append sequence to sequences currently stored in channel.
+
+        Input:
+            sequence - sequence object
+            *args    - time where sequence is called
```
[the remark this answers is not in the document]
Comment (AlexStehli) [reply qkit/measure/timedomain/TdExperiment.py]: Implemented as *times

File: qkit/measure/timedomain/TdExperiment.py [hunk 10]
```diff
@@ -0,0 +1,440 @@
...
+        """
+        if args is ():
```
[the remark this answers is not in the document]
Comment (AlexStehli) [reply qkit/measure/timedomain/TdExperiment.py]: Implemented

File: qkit/measure/timedomain/TdExperiment.py [hunk 11]
```diff
@@ -0,0 +1,440 @@
...
+        for arg in args:
+            if len0 is 0:
+                len0 = len(arg)
+            elif len(arg) is not len0:
+                print("Dimensions of input arrays do not match.")
+                return False
+        self._sequences.append(sequence)
+        self._times.append(np.vstack((args)).T)
+        if (len(self._sequences) >= 1) and self._interleave:
+            if not self.interleave():
```
[the remark this answers is not in the document]
Comment (AlexStehli) [reply qkit/measure/timedomain/TdExperiment.py]: Renamed to set_interleave

File: qkit/measure/timedomain/sequence_library.py (hunk 5 above)
Comment (AlexStehli) [reply qkit/measure/timedomain/sequence_library.py]: Does not belong in sequence_library anymore.

File: qkit/measure/timedomain/sequence_library.py (hunk 7 above)
Comment (AlexStehli) [reply qkit/measure/timedomain/sequence_library.py]: Fixed

**Comment by Schneider1** [comment]
I just scanned through the changes and found the following: 

- [ ] `measure_td`: I don't see the point in the `find_latest_file()` function. You could do it with a switch in the `measure_3D_AWG` function or store the filename internally. Scanning through all files in the data dir is not the best solution here.

- [ ] `measure_td`: `None`  comparisons should be done with `if z_unit is None:` instead of `if z_unit == None:` due to the fact that the variable could be an array. This is not acutally relevant here, but I suggest to use the `is` comparison throughout the code.
- [ ] `timedomain/initialize.py`: In my opinion, the code should still work for the old way and for different setups, so please introduce a kind of switch that allows to distinguish between different setups and to fully restore the old behavior.

**Pull request merged**